cishouseholds/impute.py

```diff
@@ -254,3 +254,393 @@ def merge_previous_imputed_values(
+            warn=True,
+        )
+
+        report_to_log(freqs.filter(col("donor_count") <= min_donors).toPandas())
+        raise ValueError(
+            "%d donor pools found with less than the required %s minimum donor(s)" % (single_donors, min_donors)
+        )
+
+    freqs = freqs.withColumn("probs", col("frequency") / col("donor_count"))
+
+    # best don uniques to get donors from
+    full_df = freqs.join(
+        imp_df.groupby("imp_uniques").agg(f.count("imp_uniques").alias("imp_group_size")), on="imp_uniques"
+    )
+
+    full_df = full_df.withColumn("exp_freq", col("probs") * col("imp_group_size"))
+
+    full_df = full_df.withColumn("int_part", floor(col("exp_freq")))
+
+    full_df = full_df.withColumn("dec_part", col("exp_freq") - col("int_part"))
+
+    ###########
+    # int part
+    ###########
+
+    # take don imp_var where int part !=0
+    full_df = full_df.withColumn("new_imp_group_size", col("imp_group_size") - col("int_part"))
+    int_dons = full_df.select("int_part", "imp_uniques", "don_" + imp_var).filter(col("int_part") >= 1)
+
+    # required decimal donors
+    dec_dons = full_df.select("probs", "imp_uniques", "don_" + imp_var, "new_imp_group_size").filter(
+        col("dec_part") > 0
+    )
+
+    ###########
+    # dec part
+    ###########
+
+    # generate random number
+    # times by prob
+    # sort by prob
+    # select number required
+    dec_dons = dec_dons.withColumn("random_number", f.rand() * col("probs"))
+    tt = Window.partitionBy("imp_uniques").orderBy(col("random_number").desc())
+    dec_dons = dec_dons.withColumn("row", row_number().over(tt)).filter(col("row") <= col("new_imp_group_size"))
+
+    # join int_dons and dec_dons
+    to_impute = int_dons.select("imp_uniques", "don_" + imp_var).unionByName(
+        dec_dons.select("imp_uniques", "don_" + imp_var)
+    )
+    ww = Window.partitionBy("imp_uniques").orderBy(col("rand"))
+    to_impute = to_impute.withColumn("rand", f.rand())
+    to_impute = to_impute.withColumn("row", row_number().over(ww))
+    to_impute = to_impute.withColumnRenamed("don_" + imp_var, imp_var)
+
+    don_df_final = df.filter((col(imp_var).isNotNull()) & ~(isnan(imp_var)))
+    xx = Window.partitionBy("imp_uniques").orderBy(col(id_column))
+    imp_df = imp_df.withColumn("row", row_number().over(xx)).drop(imp_var)
+
+    imp_df_final = imp_df.join(
+        to_impute, on=(imp_df.imp_uniques == to_impute.imp_uniques) & (imp_df.row == to_impute.row)
+    ).drop("imp_uniques", "row", "rand")
+
+    # prepare export df
+    export_df = sparkSession.createDataFrame([], df.schema)
+    export_df = export_df.unionByName(imp_df_final)
+    imputed_count = export_df.count()
+
+    report_to_log("%s records imputed." % imputed_count)
+    report_to_log("Summary statistics for imputed values: %s" % imp_var)
+    report_to_log(imp_df_final.select(imp_var).summary().toPandas())
+
+    export_df = export_df.unionByName(don_df_final)
```

Comment:
Change to outer join with imputed values. Rename imputed value column to keep separate